### PR TITLE
[ENH] Add option for channel-dependent distance computations

### PR DESCRIPTION
**Pull request by Sharathchenna**
#1992 

What does this implement/fix? Explain your changes.
Implement channel-dependent and channel-independent distance computation modes in the BaseSimilaritySearch

### CODE REVIEW

**Comment by baraline** [comment]
Hey, I would advise you to look at the similarity search notebooks to get some background on the task we're solving in this module and how the module is constructed.

The changes on dependent and independent distances must be implemented in the already existent `distance_profiles` files.

File: aeon/similarity_search/distance_profiles/naive_distance_profile.py [hunk 1]
```diff
@@ -210,18 +207,21 @@ def _normalized_naive_distance_profile(
...
                         X_stds[i_instance][i_channel, i_candidate],
                     )
-                    distance_profiles[i_instance][i_channel, i_candidate] = (
-                        numba_distance_function(q_norm[i_channel], _C)
-                    )
+
+                # Compute distance considering all channels together
+                distance_profiles[i_instance][i_candidate] = numba_distance_function(
+                    q_norm, _C
+                )
```
Comment (baraline) [inline aeon/similarity_search/distance_profiles/naive_distance_profile.py:225]: Comments from the non normalized version also apply here

**Comment by Sharathchenna** [comment]
@baraline  I made the changes you asked me to. Can you please check and let me know if there are any changes to make? Thanks!

**Comment by baraline** [comment]
Hi, having done a more in-depth review, here is what would be left to do :

- Add the channel option to the base similarity class, and pass this parameter to the function calling the distance profile in query search (`_call_distance_profile`)
- In this same function the following code :
```python
# For now, deal with the multidimensional case as "dependent", so we sum.
if self.metadata_["unequal_length"]:
    distance_profiles = List(
        [distance_profiles[i].sum(axis=0) for i in range(self.n_cases_)]
    )
else:
    distance_profiles = distance_profiles.sum(axis=1)
```
Should be deleted and moved to `euclidean_distance_profile.py` and `squared_distance_profile.py` distance profile function to implement the channel option. For example, in `euclidean_distance_profile.py`  (squared would just sum the channel in all cases):

```python
# sqrt each first and then sum : channel independent
if channel_independent and self.metadata_["unequal_length"]:
    distance_profiles = distance_profiles**0.5
    distance_profiles = List(
        [distance_profiles[i].sum(axis=0) for i in range(self.n_cases_)]
    )
elif channel_independent:
    distance_profiles = distance_profiles**0.5
    distance_profiles = distance_profiles.sum(axis=1)
# sum channels first and then sqrt : channel dependent
elif not channel_independent and self.metadata_["unequal_length"]:
    distance_profiles = List(
        [distance_profiles[i].sum(axis=0) for i in range(self.n_cases_)]
    )
    distance_profiles = distance_profiles**0.5
else:
    distance_profiles = distance_profiles.sum(axis=1)
    distance_profiles = distance_profiles**0.5
    return distance_profiles
```
This is a bit cumbersome, as we have to deal with the case where distance profiles are not arrays for unequal length inputs. (There surely is a more concise way to make this).

For the case of `naive_distance_profile`, what you did should already do the trick. 

- Add the unit test for this channel dependent / independent. 
- Modify the existing test to adapt to the changes we made at the previous step.
 
If you are not sure what to do for the test, I could edit the files myself. If you are interested in this step too, I can explain what would be expected and we can work through it together.

**Comment by Sharathchenna** [comment]
@baraline is there anything to do here? Thank You!

**Comment by baraline** [comment]
Hey, sent you message on Slack, busy preparing for ECML conference next week. 
But we'll need to cover testing together after that i'll try to review it this week but can't guarantee it

**Comment by baraline** [comment]
Hey, there will be a bit of adapting to the new content added recently before we go on the testing part of it. 

You can update your branch from origin main branch and once this is done we'll talk about the testing part together!

**Comment by Sharathchenna** [comment]
@baraline Hey, I have updated my branch from main. Can you please tell me what to do now?

**Comment by baraline** [comment]
You should first resolve the conflicts (and push the code here). Once this is done we can move to testing

**Comment by Sharathchenna** [comment]
> You should first resolve the conflicts (and push the code here). Once this is done we can move to testing

Hey there! I’ve sorted out the conflicts and fixed the code.

**Comment by Sharathchenna** [comment]
@baraline 
Hey Hi! I’ve sorted out the conflicts and fixed the code.

**Comment by baraline** [comment]
Hi @Sharathchenna, I appreciate your patience with this. Unfortunately, I think I was a bit too optimistic when creating the issue. I made some tries at implementing new similarity search algorithm last week, and the channel dependent option we are trying to implement wouldn't allow these methods to perform properly.

As I think it was only a feature request from me (i.e. not motivated by a real user need), I think we should stop the development of this functionality. I'm really sorry for the time you have invested in this, and hope you understand the decision.

**Comment by Sharathchenna** [comment]
> Hi @Sharathchenna, I appreciate your patience with this. Unfortunately, I think I was a bit too optimistic when creating the issue. I made some tries at implementing new similarity search algorithm last week, and the channel dependent option we are trying to implement wouldn't allow these methods to perform properly.
> 
> As I think it was only a feature request from me (i.e. not motivated by a real user need), I think we should stop the development of this functionality. I'm really sorry for the time you have invested in this, and hope you understand the decision.

Hey there! No worries at all, should I close this issue? Thanks a bunch for all your help! :)

**Comment by baraline** [comment]
I'll do it don't worry, thanks !

**Pull request closed without merging**